src/pip/_internal/operations/install/wheel.py

```diff
@@ -575,9 +576,8 @@ def is_entrypoint_wrapper(name):
 
     # Record details of all files installed
     record = os.path.join(dest_info_dir, 'RECORD')
-    temp_record = os.path.join(dest_info_dir, 'RECORD.pip')
     with open_for_csv(record, 'r') as record_in:
-        with open_for_csv(temp_record, 'w+') as record_out:
+        with adjacent_tmp_file(record, mode='w') as record_out:
```

Comment:
Any particular reason to change `w+` to `w`?

Reply:
Also, are we sure that `adjacent_tmp_file` is equivalent to `open_for_csv` for our purpose here? (I've not investigated myself yet)

Reply:
> Any particular reason to change `w+` to `w`?

We are not reading anything from this file so why request the extra permissions?

Reply:
> Also, are we sure that `adjacent_tmp_file` is equivalent to `open_for_csv` for our purpose here? (I've not investigated myself yet)

All `open_for_csv` does is append some parameters to `open()` to account for differences between different versions of Python.

`adjacent_tmp_file` calls `NamedTemporaryFile` which in turn calls `_sanitize_params`, calls `close()` on exit (in addition to that of `file`) and ensures file is written to disk before it's closed.

Tests would not pass if binary mode was used even with Python 2 but they do otherwise (on both versions). So I do not see any obstacles to replacing `open_for_csv`.

Reply:
Getting the open parameters correct for CSV files is fairly subtle, and having `open_for_csv` is important to make sure it's easy for people to get it correct even if they aren't aware of platform differences. Having text mode and `newline=''` is crucial for correct handling on Windows under Python 3 (and having binary mode is crucial on Python 2). So I'd be concerned that the switch to `adjacent_tmp_file` is generating a CSV file with incorrect line endings. Do we know for sure that we have a test that checks this?

Note: I don't actually know (without doing some research) the precise "rules"<sup>1</sup> for the format - but I *do* know that the docs in the CSV module (in particular the footnote [here](https://docs.python.org/3/library/csv.html#id3)) were arrived at after quite a lot of frustrating investigation into weird bugs.

<sup>1</sup> CSV is not a particularly well defined format, so they aren't actually so much rules as "hard-won knowledge about what works" 🙄

Reply:
This code is definitely passed by tests as they first failed when binary mode was used. This happened for both versions of Python because the CSV writer deals with encoding inside and insists on writing strings no matter whether string and byte data is passed in the argument. I am not entirely quite sure why binary mode worked with Python 2 under `open_for_csv`. Nevertheless, the tests covered the code and passed.

As for newline, `adjacent_temp_file` eventually uses the `io` module the [documentation](https://docs.python.org/2/library/io.html) for which states:

> On output, if newline is None, any '\n' characters written are translated to the system default line separator, os.linesep. If newline is '', no translation takes place. If newline is any of the other legal values, any '\n' characters written are translated to the given string.

Again, based on the tests passing for all platforms and versions, it seems logical to conclude that the mode and newlines are handled appropriately.

Reply:
But the CSV module docs that I referenced, explicitly state that on Python 3, you must not use binary and you must use `newline=''`, and on Python 2 you must use binary mode.

I'm a strong -1 on implementing something that deliberately violates the requirements documented in the CSV module - even if our tests imply that doing so is OK (I'd question the tests in that case).

Reply:
I see the link to the newline comment and this is addressed by the `io` module internally. I do not see where it says Python 2 insists on binary mode. A simple test shows:
```
Python 2.7.10 (default, Feb 22 2019, 21:17:52) 
[GCC 4.2.1 Compatible Apple LLVM 10.0.1 (clang-1001.0.37.14)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import csv
>>> with open('/tmp/bla', 'w') as f:
...     c = csv.writer(f)
...     c.writerow(['hello'])
... 
>>> with open('/tmp/bla', 'r') as f:
...     c = csv.reader(f)
...     print '\n'.join(','.join(l) for l in c)
... 
hello
>>> 
```

Binary also works:
```
>>> with open('/tmp/bla', 'rb') as f:
...     c = csv.reader(f)
...     print '\n'.join(','.join(l) for l in c)
... 
hello
```

Reply:
As per [documentation](https://docs.python.org/2/library/functions.html#open) for `open()`:

> Thus, when opening a binary file, you should append 'b' to the mode value to open the file in binary mode, which will improve portability. (Appending 'b' is useful even on systems that don’t treat binary and text files differently, where it serves as documentation.) See below for more possible values of mode.

Since we only ever deal with text data in CSVs there is no reason why it should matter. All Python documentation uses `b` everywhere because in Python 2 it essentially had no adverse effect. It was required for some systems and harmless for others.

Reply:
https://docs.python.org/2/library/csv.html#csv.reader - "If csvfile is a file object, it must be opened with the ‘b’ flag on platforms where that makes a difference". You're running on MacOS, which is a platform where it *doesn't* make a difference. SO it's not surprising you can't demonstrate an issue (your specific test *does* work on Windows, but again, that just means I suspect your test is incomplete - I have had actual, real-life errors caused by failing to use binary mode on Windows).

Anyhow, regardless of how much we debate test cases that demonstrate one thing or another, that's irrelevant, We should follow the documented requirements of the CSV module, not waste time debating whether they are needed. If you feel sufficiently strongly, by all means raise a CPython bug that claims the requirements in the docs are obsolete and should be removed.

Reply:
So then I can extract the `version -> mode` part from `open_for_csv` and call it there in addition to `adjacent_tmp_file(file, 'w' + get_mode())`.

Reply:
For what it's worth I did [this](https://github.com/pypa/pip/pull/7651/files#diff-b16e00f0552db1dbd9a998e25a075c92R234) in our recent wheel builder helper test code, which was the easiest way I could find that was compatible.

This would also avoid needing to make changes to `adhwcent_tmp_file`.